# Task 1. SVC_WINEQUALITY

In [1]:
import pandas as pd
url = 'https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv'
data=pd.read_csv(url)

C:\Users\lglv\Anaconda3\lib\site-packages\pandas\compat\_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)


In [2]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [5]:
data.describe().T

count       mean        std      min      25%  \
fixed acidity         1599.0   8.319637   1.741096  4.60000   7.1000   
volatile acidity      1599.0   0.527821   0.179060  0.12000   0.3900   
citric acid           1599.0   0.270976   0.194801  0.00000   0.0900   
residual sugar        1599.0   2.538806   1.409928  0.90000   1.9000   
chlorides             1599.0   0.087467   0.047065  0.01200   0.0700   
free sulfur dioxide   1599.0  15.874922  10.460157  1.00000   7.0000   
total sulfur dioxide  1599.0  46.467792  32.895324  6.00000  22.0000   
density               1599.0   0.996747   0.001887  0.99007   0.9956   
pH                    1599.0   3.311113   0.154386  2.74000   3.2100   
sulphates             1599.0   0.658149   0.169507  0.33000   0.5500   
alcohol               1599.0  10.422983   1.065668  8.40000   9.5000   
quality               1599.0   5.636023   0.807569  3.00000   5.0000   

                           50%        75%        max  
fixed acidity          7.90000   9.200000   15.90000  
volatile acidity       0.52000   0.640000    1.58000  
citric acid            0.26000   0.420000    1.00000  
residual sugar         2.20000   2.600000   15.50000  
chlorides              0.07900   0.090000    0.61100  
free sulfur dioxide   14.00000  21.000000   72.00000  
total sulfur dioxide  38.00000  62.000000  289.00000  
density                0.99675   0.997835    1.00369  
pH                     3.31000   3.400000    4.01000  
sulphates              0.62000   0.730000    2.00000  
alcohol               10.20000  11.100000   14.90000  
quality                6.00000   6.000000    8.00000

In [6]:
data.quality

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

In [7]:
data['quality'].value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [8]:
data.quality.unique()

array([5, 6, 7, 4, 8, 3], dtype=int64)

In [9]:
X = data.iloc[:,:-1]
X

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  
0         9.4  
1         9.8  
2         9.8  
3         9.8  
4         9.4  
...       ...  
1594     10.5  
1595     11.2  
1596     11.0  
1597     10.2  
1598     11.0  

[1599 rows x 11 columns]

In [10]:
y = data['quality']
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

## Splitting into: train and test sets

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(1071, 11)
(1071,)
(528, 11)
(528,)


## Scaling the data train

In [12]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train) # Calculate the mean of (sd) std dev

StandardScaler()

In [13]:
print(scaler.mean_)

[ 8.30345472  0.53246499  0.26933707  2.54691877  0.08772736 15.91223156
 46.76330532  0.99677933  3.31453782  0.65881419 10.41521942]


In [15]:
X_train_tr = scaler.transform(X_train)
X_train_tr

array([[ 2.40069523, -1.03103722,  1.12742595, ..., -1.26096312,
         0.52726134, -0.01431863],
       [-0.93967131,  1.22920403, -1.32502245, ...,  1.52622836,
        -0.28225704,  2.24363201],
       [-0.99827424,  0.55113165, -1.37611513, ..., -0.74241587,
        -1.20742091, -0.86105011],
       ...,
       [-0.6466567 ,  0.49462562, -1.06955908, ...,  1.26695473,
        -0.68701624, -0.86105011],
       [-0.23643625, -1.87862768,  0.4121285 , ...,  0.03540501,
         0.81637505,  1.39690052],
       [-1.46709761, -1.3700734 , -0.04770558, ...,  0.48913386,
        -0.68701624,  2.90220094]])

In [16]:
X_test_tr = scaler.transform(X_test)
X_test_tr

array([[-3.53642095e-01,  1.55589436e-01, -9.67373729e-01, ...,
        -4.83142240e-01,  6.85666499e-03, -7.66968836e-01],
       [-2.95039173e-01, -1.83446751e-01, -5.07539654e-01, ...,
         4.89133857e-01, -1.03395269e+00, -8.61050113e-01],
       [ 1.40444556e+00,  7.77155778e-01, -2.52076279e-01, ...,
        -2.23868614e-01,  1.85718440e+00, -4.84725007e-01],
       ...,
       [-2.02456406e-03, -1.25706134e+00,  6.16499196e-01, ...,
        -2.94133945e-02,  6.42906824e-01,  1.96138818e+00],
       [-6.06274859e-02,  4.50655383e+00, -1.37611513e+00, ...,
         1.39659155e+00, -9.76129945e-01,  4.56087756e-01],
       [ 4.66798811e-01,  7.20649747e-01, -6.09725004e-01, ...,
        -2.23868614e-01, -6.87016236e-01, -7.66968836e-01]])

## SVM model

In [17]:
from sklearn.svm import SVC
model = SVC()
model.fit(X_train_tr, y_train)

SVC()

In [18]:
model.score(X_train_tr, y_train)

0.6778711484593838

In [20]:
y_pred = model.predict(X_test_tr)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.5984848484848485

## Logistic Regression model

In [21]:
from sklearn.linear_model import LogisticRegression
model2 = LogisticRegression()
model2.fit(X_train_tr, y_train)

LogisticRegression()

In [22]:
model2.score(X_train_tr, y_train)

0.6218487394957983

In [23]:
y_pred_2 = model2.predict(X_test_tr)
accuracy_score(y_test, y_pred_2)

0.571969696969697

In [24]:
X_test_tr[0]

array([-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
       -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
       -0.76696884])

In [25]:
model2.predict([X_test_tr[0]])

array([5], dtype=int64)

## Randomized Search CV

In [26]:
import numpy as np
from sklearn.model_selection import RandomizedSearchCV
parameters = {'kernel':('linear', 'rbf'), 'C':[1, 10]}

In [27]:
from sklearn import svm
svm_cls=svm.SVC()
svm_cls=RandomizedSearchCV(estimator=svm_cls, param_distributions=parameters,n_iter=100,cv=5,verbose=2,
                               random_state=100,n_jobs=-1)
### fit the randomized model
svm_cls.fit(X_train_tr, y_train)

C:\Users\lglv\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:289: UserWarning: The total space of parameters 4 is smaller than n_iter=100. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  % (grid_size, self.n_iter, grid_size), UserWarning)


Fitting 5 folds for each of 4 candidates, totalling 20 fits


RandomizedSearchCV(cv=5, estimator=SVC(), n_iter=100, n_jobs=-1,
                   param_distributions={'C': [1, 10],
                                        'kernel': ('linear', 'rbf')},
                   random_state=100, verbose=2)

In [28]:
svm_cls.best_params_

{'kernel': 'rbf', 'C': 10}

In [29]:
best_SVM_grid=svm_cls.best_estimator_

In [30]:
y_pred_RG = best_SVM_grid.predict(X_test_tr)
accuracy_score(y_test, y_pred_RG)

0.615530303030303

## GridSearch CV

In [31]:
svm_cls.best_params_

{'kernel': 'rbf', 'C': 10}

In [32]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'kernel': [svm_cls.best_params_['kernel']],
    'C': [svm_cls.best_params_['C']-4, 
          svm_cls.best_params_['C']-2,
           svm_cls.best_params_['C'],
           svm_cls.best_params_['C']+2,
           svm_cls.best_params_['C']+4]
}

print(param_grid)

{'kernel': ['rbf'], 'C': [6, 8, 10, 12, 14]}


In [33]:
svm_cls_gs=svm.SVC()
grid_search=GridSearchCV(estimator=svm_cls_gs,param_grid=param_grid,cv=10,n_jobs=-1,verbose=2)
grid_search.fit(X_train_tr, y_train)

Fitting 10 folds for each of 5 candidates, totalling 50 fits


C:\Users\lglv\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:668: UserWarning: The least populated class in y has only 8 members, which is less than n_splits=10.
  % (min_groups, self.n_splits)), UserWarning)


GridSearchCV(cv=10, estimator=SVC(), n_jobs=-1,
             param_grid={'C': [6, 8, 10, 12, 14], 'kernel': ['rbf']},
             verbose=2)

In [34]:
grid_search.best_estimator_

SVC(C=8)

In [36]:
best_grid=grid_search.best_estimator_
best_grid

SVC(C=8)

In [38]:
y_pred_GS = best_grid.predict(X_test_tr)
accuracy_score(y_test, y_pred_GS)

0.6098484848484849

## Conclusions for task 1

### Randomized Search CV gave the best score:  0.615
### Grid Search CV gave the second best score: 0.609

# Task 2: Graduate Admission Prediction

In [39]:
url = 'https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv'
df = pd.read_csv(url)
df

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0             1        337          118                  4  4.5   4.5  9.65   
1             2        324          107                  4  4.0   4.5  8.87   
2             3        316          104                  3  3.0   3.5  8.00   
3             4        322          110                  3  3.5   2.5  8.67   
4             5        314          103                  2  2.0   3.0  8.21   
..          ...        ...          ...                ...  ...   ...   ...   
495         496        332          108                  5  4.5   4.0  9.02   
496         497        337          117                  5  5.0   5.0  9.87   
497         498        330          120                  5  4.5   5.0  9.56   
498         499        312          103                  4  4.0   5.0  8.43   
499         500        327          113                  4  4.5   4.5  9.04   

     Research  Chance of Admit   
0           1              0.92  
1           1              0.76  
2           1              0.72  
3           1              0.80  
4           0              0.65  
..        ...               ...  
495         1              0.87  
496         1              0.96  
497         1              0.93  
498         0              0.73  
499         0              0.84  

[500 rows x 9 columns]

In [40]:
df.describe().T

count       mean         std     min       25%     50%  \
Serial No.         500.0  250.50000  144.481833    1.00  125.7500  250.50   
GRE Score          500.0  316.47200   11.295148  290.00  308.0000  317.00   
TOEFL Score        500.0  107.19200    6.081868   92.00  103.0000  107.00   
University Rating  500.0    3.11400    1.143512    1.00    2.0000    3.00   
SOP                500.0    3.37400    0.991004    1.00    2.5000    3.50   
LOR                500.0    3.48400    0.925450    1.00    3.0000    3.50   
CGPA               500.0    8.57644    0.604813    6.80    8.1275    8.56   
Research           500.0    0.56000    0.496884    0.00    0.0000    1.00   
Chance of Admit    500.0    0.72174    0.141140    0.34    0.6300    0.72   

                      75%     max  
Serial No.         375.25  500.00  
GRE Score          325.00  340.00  
TOEFL Score        112.00  120.00  
University Rating    4.00    5.00  
SOP                  4.00    5.00  
LOR                  4.00    5.00  
CGPA                 9.04    9.92  
Research             1.00    1.00  
Chance of Admit      0.82    0.97

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Serial No.         500 non-null    int64  
 1   GRE Score          500 non-null    int64  
 2   TOEFL Score        500 non-null    int64  
 3   University Rating  500 non-null    int64  
 4   SOP                500 non-null    float64
 5   LOR                500 non-null    float64
 6   CGPA               500 non-null    float64
 7   Research           500 non-null    int64  
 8   Chance of Admit    500 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 35.3 KB


## Renaming columns

In [42]:
df.rename(columns = {'Serial No.':'Serial_No.',
                     'GRE Score':'GRE_Score',
                     'TOEFL Score':'TOEFL_Score',
                     'University Rating':'University_Rating',
                    'Chance of Admit ':'Chance_of_Admit'}, inplace = True)

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Serial_No.         500 non-null    int64  
 1   GRE_Score          500 non-null    int64  
 2   TOEFL_Score        500 non-null    int64  
 3   University_Rating  500 non-null    int64  
 4   SOP                500 non-null    float64
 5   LOR                500 non-null    float64
 6   CGPA               500 non-null    float64
 7   Research           500 non-null    int64  
 8   Chance_of_Admit    500 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 35.3 KB


In [44]:
df.columns

Index(['Serial_No.', 'GRE_Score', 'TOEFL_Score', 'University_Rating', 'SOP',
       'LOR ', 'CGPA', 'Research', 'Chance_of_Admit'],
      dtype='object')

In [47]:
X_adm = df.iloc[:,1:-1] # 'Serial_No.' feature is not considered in the X data set
X_adm.head()

GRE_Score  TOEFL_Score  University_Rating  SOP  LOR   CGPA  Research
0        337          118                  4  4.5   4.5  9.65         1
1        324          107                  4  4.0   4.5  8.87         1
2        316          104                  3  3.0   3.5  8.00         1
3        322          110                  3  3.5   2.5  8.67         1
4        314          103                  2  2.0   3.0  8.21         0

In [48]:
y_adm = df['Chance_of_Admit']
y_adm.head()

0    0.92
1    0.76
2    0.72
3    0.80
4    0.65
Name: Chance_of_Admit, dtype: float64

## Splitting into: train and test sets

In [49]:
from sklearn.model_selection import train_test_split
X_train_adm, X_test_adm, y_train_adm, y_test_adm = train_test_split(X_adm, y_adm, test_size=0.33, random_state=42)
print(X_train_adm.shape)
print(y_train_adm.shape)
print(X_test_adm.shape)
print(y_test_adm.shape)

(335, 7)
(335,)
(165, 7)
(165,)


## Scaling the data train

In [50]:
scaler_adm = StandardScaler()
X_train_adm_ss = scaler_adm.fit_transform(X_train_adm)
X_train_adm_ss

array([[-5.25174731e-01, -7.06985745e-01, -8.28848674e-02, ...,
         5.38819022e-01, -8.35765678e-01, -1.14470294e+00],
       [ 4.77035981e-01,  2.76433873e-01,  1.65251704e+00, ...,
        -1.61323061e-03,  3.63045482e-01, -1.14470294e+00],
       [-7.98504925e-01, -1.03479228e+00, -9.50585823e-01, ...,
         5.38819022e-01, -1.15657430e+00, -1.14470294e+00],
       ...,
       [-1.34516531e+00, -1.36259882e+00, -1.81828678e+00, ...,
        -1.62290999e+00, -2.25407747e+00, -1.14470294e+00],
       [-7.07394861e-01, -3.79179206e-01, -9.50585823e-01, ...,
         5.38819022e-01, -1.52803691e+00, -1.14470294e+00],
       [-2.51844537e-01, -2.15275936e-01, -9.50585823e-01, ...,
        -1.61323061e-03, -5.65611050e-01, -1.14470294e+00]])

In [51]:
X_test_adm_ss = scaler_adm.transform(X_test_adm)
X_test_adm_ss

array([[ 1.57035676e+00,  1.42375676e+00,  7.84816088e-01, ...,
        -1.61323061e-03,  1.61251063e+00,  8.73589088e-01],
       [-2.51844537e-01,  1.12530603e-01,  7.84816088e-01, ...,
         5.38819022e-01,  7.68277423e-01,  8.73589088e-01],
       [-1.60734472e-01, -3.79179206e-01, -9.50585823e-01, ...,
        -1.08247774e+00, -1.57869090e+00, -1.14470294e+00],
       ...,
       [-6.96244077e-02, -7.06985745e-01, -9.50585823e-01, ...,
         1.07925128e+00,  2.61737496e-01, -1.14470294e+00],
       [-6.16284796e-01, -1.36259882e+00, -9.50585823e-01, ...,
        -1.62290999e+00, -2.16965415e+00, -1.14470294e+00],
       [ 2.14856571e-02, -2.15275936e-01, -1.81828678e+00, ...,
        -1.61323061e-03, -1.57869090e+00,  8.73589088e-01]])

## Support Vector Regression (SVR) model

In [52]:
from sklearn.svm import SVR
model_adm = SVR()
model_adm.fit(X_train_adm_ss, y_train_adm)

SVR()

In [53]:
model_adm.score(X_train_adm_ss, y_train_adm)

0.7757519357301843

In [58]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
y_pred_adm = model_adm.predict(X_test_adm_ss)
print('r2:', r2_score(y_test_adm, y_pred_adm))
print('MAE:', mean_absolute_error(y_test_adm, y_pred_adm))
print('MSE:',mean_squared_error(y_test_adm, y_pred_adm))

r2: 0.7601877598637249
MAE: 0.05758218406189208
MSE: 0.004940947816722295


## Preliminary conclusions for task 2
### In this point

*  The training accuracy is: 0.77
*  The prediction accuracy is: 0.76

#TASK

1. YOU HAVE TO INCREASE THE ACCURACY OF THE SVC MODEL(WINEQUALITY DATASET)
2. HYPERPARAMETER TUNING(GRIDSEARCH cv https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
3. YOU HVAE TO IMPLEMENT SVR(ADDIMISSOIN_PREDICTION)
https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv

https://docs.google.com/forms/d/e/1FAIpQLSeCob4z7rrWCy86YKAHFXjUGMvV6H40wXVlY8jdXnqNSvGiTQ/viewform